storagemarket/impl/client.go

```diff
@@ -372,4 +369,19 @@ func (c *clientDealEnvironment) CloseStream(proposalCid cid.Cid) error {
 	return c.c.conns.Disconnect(proposalCid)
 }
 
-var _ clientstates.ClientDealEnvironment = &clientDealEnvironment{}
+func (c *clientDealEnvironment) StartDataTransfer(ctx context.Context, to peer.ID, voucher datatransfer.Voucher, baseCid cid.Cid, selector ipld.Node) error {
+	_, err := c.c.dataTransfer.OpenPushDataChannel(ctx, to, voucher, baseCid, selector)
+	return err
+}
+
+func (c *clientDealEnvironment) OnDataTransferEvent(ctx context.Context, eventCode datatransfer.EventCode, cb providerstates.OnDataTransferEventCallback) datatransfer.Unsubscribe {
+	// Can there be overlapping Open/Complete events?
+	return c.c.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
+		if event.Code == eventCode {
+			err := cb(event)
```

Comment:
SubscribeToEvents will get called for every transfer possible. I don't think this is a good way to setup this listener.

You could use the approach above and add a Proposal CID and check the Voucher (also check that it's a data transfer voucher

OR

You could use a single subscriber, check the voucher, and send events in from the outside -- the same way as was used with DataTransferSubscriber on the provider.

I think this is more efficient.

And then you just avoid a StateHandler entirely for StorageDealTransferring.

Reply:
That makes way more sense.  I was thinking of the "resuming" scenario, but this covers all the cases.

Reply:
Oh right.